File: sdk/typescript/src/vault.ts
import { PublicKey } from '@solana/web3.js';
import { Algorithm, QuantumVaultAccount } from './types';

export class QuantumVault {
  public address: PublicKey;
  public owner: PublicKey;
  public algorithm: Algorithm;
  public publicKey: Buffer;
  public createdAt: Date;
  public lastKeyRotation: Date;
  public transactionCount: number;

  constructor(account: QuantumVaultAccount, address: PublicKey) {
    this.address = address;
    this.owner = account.owner;
    this.algorithm = account.algorithm;
    this.publicKey = Buffer.from(account.publicKey);
    this.createdAt = new Date(account.createdAt.toNumber() * 1000);
    this.lastKeyRotation = new Date(account.lastKeyRotation.toNumber() * 1000);
    this.transactionCount = account.transactionCount.toNumber();
  }

  isKeyRotationRequired(): boolean {
    const now = Date.now();
    const rotationPeriod = 90 * 24 * 60 * 60 * 1000;
    return now - this.lastKeyRotation.getTime() > rotationPeriod;
  }

  getAlgorithmName(): string {
    return Algorithm[this.algorithm];
  }
}

File: sdk/rust/src/vault.rs
//! Quantum-safe vault management

use crate::{Result, crypto::Algorithm};
use solana_sdk::pubkey::Pubkey;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QuantumVault {
    pub address: Pubkey,
    pub owner: Pubkey,
    pub algorithm: Algorithm,
    pub public_key: Vec<u8>,
    pub created_at: i64,
    pub last_key_rotation: i64,
    pub transaction_count: u64,
}

impl QuantumVault {
    pub fn is_key_rotation_required(&self) -> bool {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64;
        const ROTATION_PERIOD: i64 = 90 * 24 * 60 * 60;
        now - self.last_key_rotation > ROTATION_PERIOD
    }
}
